components/tidb_query/src/rpn_expr/impl_cast.rs

```diff
@@ -304,6 +338,148 @@ pub fn cast_uint_as_string(val: &Option<Int>) -> Result<Option<Bytes>> {
+#[inline]
+pub fn cast_string_as_json(
+    extra: &RpnFnCallExtra<'_>,
+    val: &Option<Bytes>,
+) -> Result<Option<Json>> {
+    match val {
+        None => Ok(None),
+        Some(val) => {
+            if extra
+                .ret_field_type
+                .flag()
+                .contains(FieldTypeFlag::PARSE_TO_JSON)
+            {
+                let s = box_try!(String::from_utf8(val.to_owned()));
+                let val: Json = s.parse()?;
+                Ok(Some(val))
+            } else {
+                // FIXME: port `JSONBinary` from TiDB to adapt if the bytes is not a valid utf8 string
+                let val = unsafe { String::from_utf8_unchecked(val.to_owned()) };
+                Ok(Some(Json::String(val)))
+            }
+        }
+    }
+}
+
+/// The implementation for push down signature `CastIntAsDuration`
+#[rpn_fn(capture = [ctx, extra])]
+#[inline]
+pub fn cast_int_as_duration(
+    ctx: &mut EvalContext,
+    extra: &RpnFnCallExtra<'_>,
+    val: &Option<Int>,
+) -> Result<Option<Duration>> {
+    match val {
+        None => Ok(None),
+        Some(val) => {
+            let dur = Duration::from_i64(ctx, *val, extra.ret_field_type.get_decimal() as u8)?;
+            Ok(Some(dur))
+        }
+    }
+}
+
+#[rpn_fn(capture = [extra])]
+#[inline]
+pub fn cast_real_as_duration(
+    extra: &RpnFnCallExtra<'_>,
+    val: &Option<Real>,
+) -> Result<Option<Duration>> {
+    match val {
+        None => Ok(None),
+        Some(val) => {
+            // FIXME: the `Duration::parse` should return `Res` which indicates whether `Overflow` or `Truncated` occurred.
+            let dur = Duration::parse(
+                val.into_inner().to_string().as_bytes(),
+                extra.ret_field_type.get_decimal() as i8,
+            )?;
+            Ok(Some(dur))
+        }
+    }
+}
+
+#[rpn_fn(capture = [extra])]
+#[inline]
+pub fn cast_bytes_as_duration(
+    extra: &RpnFnCallExtra<'_>,
+    val: &Option<Json>,
+) -> Result<Option<Duration>> {
+    match val {
+        None => Ok(None),
+        Some(val) => {
+            let s = val.unquote()?;
+            let dur = Duration::parse(s.as_bytes(), extra.ret_field_type.get_decimal() as i8)?;
+            Ok(Some(dur))
+        }
+    }
+}
+
+#[rpn_fn(capture = [extra])]
+#[inline]
+pub fn cast_decimal_as_duration(
+    extra: &RpnFnCallExtra<'_>,
+    val: &Option<Decimal>,
+) -> Result<Option<Duration>> {
+    match val {
+        None => Ok(None),
+        Some(val) => {
+            let dur = Duration::parse(
+                val.to_string().as_bytes(),
+                extra.ret_field_type.get_decimal() as i8,
+            )?;
+            Ok(Some(dur))
+        }
+    }
+}
+
+#[rpn_fn(capture = [extra])]
+#[inline]
+pub fn cast_json_as_duration(
+    extra: &RpnFnCallExtra<'_>,
+    val: &Option<Decimal>,
+) -> Result<Option<Duration>> {
+    match val {
+        None => Ok(None),
+        Some(val) => {
+            let dur = Duration::parse(
```

Comment:
How about changing ConvertTo to include target_field_type?

Reply:
It's not the time to introduce this parameter for `ConvertTo` trait (it may introduce a huge of changes). Anyway, this is the last PR for RPN cast functions (it's not a high priority issue).